csrc/multidevice/utils.cpp

```diff
@@ -254,7 +257,13 @@ void insertReshardings(Fusion* fusion) {
   // Remove this after we refactor this as a pre-segmenter pass.
   FusionGuard fg(fusion);
   auto exprs = fusion->exprs();
+  // Replacing expression inputs creates a new expression. Map the old
+  // expression to new expression, so that we can continue iterating.
+  std::unordered_map<Expr*, Expr*> replacement_map = {};
   for (auto expr : exprs) {
+    if (replacement_map.find(expr) != replacement_map.end()) {
+      expr = replacement_map[expr];
```

Comment:
I understand the need for this. However, correct me if Im wrong, but I'm afraid there could be cases with multiple replacements composed, right?
e.g. `expr0` -> `expr1` -> `expr2` if the pass changes both the inputs and outputs of the expr. In this case, we would need here to apply the map in a while loop.

Otherwise, maybe it is simpler/more general to make `insertReshardings` recursive

Reply:
Yes, there is definitely an issue. I don't quite see the one you have mentioned, but I do see this one:
```
tv0=expr0(tv)
tv1=expr1(tv)
tv2 = expr2(tv0, tv1)
```

If all these expressions are resharding and require sets, then the heuristic would insert a set after expr0 for tv0, after expr1 for tv1, and before expr2 for tv0/tv1.

If we 'shard after' expr0 and expr1 to form tv0' and tv1' then both iterations would replace expr2 with`ir_utils::replaceValInAllExprInputsAndFusionOutputs` but`replacement_map` will map the original expr2 to the first replacement.

Ideally I would like to just mutate expr2 instead of replace it, because tv0' and tv1' will be identical to the original tv0 and tv1. 
@wujingyue maybe there is a better way to do this kind of mutation?

Reply:
Instead of using `replacement_map`, can we do replacement in two passes? The first pass only replaces the inputs and therefore can be done in topological order as in `fusion->exprs()`. This is the same as before this PR. The second pass only replaces the outputs and therefore can be done in the reverse order of `fusion->exprs()`.

Reply:
This worked! I updated all the passes that changed downstream operations to iterate through fusion->exprs() backwards, so no more tracking replaced exprs.